File: src/main/java/pocket/transpiler/Transpiler.java
package pocket.transpiler;

import org.jetbrains.annotations.NotNull;
import pocket.ast.ModuleFnNameGenerator;
import pocket.ast.Program;

public interface Transpiler {
  /**
   * Transpiles a Pocket abstract syntax tree into a target language.
   *
   * @param entryFilepath The path to the Pocket source file.
   * @param program A Pocket abstract syntax tree.
   * @return A target code string.
   */
  @NotNull
  String transpile(@NotNull final String entryFilepath, @NotNull final Program program);

  @NotNull
  ModuleFnNameGenerator getModuleFnNameGenerator();
}

File: src/main/java/pocket/transpiler/Transpiler.kt
package pocket.transpiler

import pocket.ast.node.Program

/**
 * Represents a transpiler that transpiles a Pocket program abstract syntax tree
 * into a target language program.
 *
 * @see pocket.ast.node.Program
 */
abstract class Transpiler() {
    abstract fun transpile(program: Program): String
}